src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -78,22 +80,54 @@ internal class CompletionHandler : IRequestHandler<LSP.CompletionParams, LSP.Com
             var completionTrigger = await ProtocolConversions.LSPToRoslynCompletionTriggerAsync(request.Context, document, position, cancellationToken).ConfigureAwait(false);
 
             var list = await completionService.GetCompletionsAsync(document, position, completionTrigger, options: completionOptions, cancellationToken: cancellationToken).ConfigureAwait(false);
-            if (list == null)
+            if (list == null || list.Items.IsEmpty)
             {
                 return null;
             }
 
             var lspVSClientCapability = context.ClientCapabilities?.HasVisualStudioLspCapability() == true;
-
             var commitCharactersRuleCache = new Dictionary<ImmutableArray<CharacterSetModificationRule>, ImmutableArray<string>>();
 
             // Cache the completion list so we can avoid recomputation in the resolve handler
             var resultId = await _completionListCache.UpdateCacheAsync(list, cancellationToken).ConfigureAwait(false);
 
+            // Feature flag to enable the return of TextEdits instead of InsertTexts (will increase payload size).
+            // Flag is defined in VisualStudio\Core\Def\PackageRegistration.pkgdef.
+            // We also check against the CompletionOption for test purposes only.
+            Contract.ThrowIfNull(context.Solution);
+            var featureFlagService = context.Solution.Workspace.Services.GetRequiredService<IExperimentationService>();
+            var returnTextEdits = featureFlagService.IsExperimentEnabled(WellKnownExperimentNames.LSPCompletion) ||
+                completionOptions.GetOption(CompletionOptions.ForceRoslynLSPCompletionExperiment, document.Project.Language);
+
+            SourceText? documentText = null;
+            TextSpan? defaultSpan = null;
+            LSP.Range? defaultRange = null;
+            if (returnTextEdits)
+            {
+                // We want to compute the document's text just once.
+                documentText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+                var completionChange = await completionService.GetChangeAsync(
+                    document, list.Items.First(), cancellationToken: cancellationToken).ConfigureAwait(false);
```

Comment:
Given you're doing `list.Items.First()` here it sounds like you're warming up some sort of cache and seems a bit like a hack. Probably worth a comment assuming i'm right. I guess in your case this probably just represents the applicableToSpan which sounds like is always equivalent no matter the completion item?